crates/sui-edge-proxy/src/handlers.rs

```diff
@@ -0,0 +1,200 @@
+}
+
+pub async fn proxy_handler(
+    State(state): State<AppState>,
+    request: Request<Body>,
+) -> Result<Response, (StatusCode, String)> {
+    let (parts, body) = request.into_parts();
+    if parts.headers.get("content-type")
+        != Some(&reqwest::header::HeaderValue::from_static(
+            "application/json",
+        ))
+    {
+        info!("Content type is not application/json");
+        return Ok(Response::builder()
+            .status(StatusCode::BAD_REQUEST)
+            .body(Body::from("Content type must be application/json"))
+            .unwrap());
+    }
+    let body_bytes = match axum::body::to_bytes(body, usize::MAX).await {
+        Ok(bytes) => bytes,
+        Err(e) => {
+            warn!("Failed to read request body: {}", e);
+            return Ok(Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(Body::from("Failed to read request body"))
+                .unwrap());
+        }
+    };
+
+    match parts
+        .headers
+        .get("Client-Request-Method")
+        .and_then(|h| h.to_str().ok())
+    {
+        Some("sui_executeTransactionBlock") => {
+            info!("Using execution peer");
+            // no need to check the request body, skip right to proxying to execution peer
+            proxy_request(state, parts, body_bytes, true).await
+        }
+        _ => {
+            let json_body = match serde_json::from_slice::<serde_json::Value>(&body_bytes) {
+                Ok(json_body) => json_body,
+                Err(_) => {
+                    return Ok(Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(Body::from("Request body is not valid JSON"))
+                        .unwrap());
+                }
+            };
+            if let Some("sui_executeTransactionBlock") =
+                json_body.get("method").and_then(|m| m.as_str())
+            {
+                proxy_request(state, parts, body_bytes, true).await
+            } else {
+                proxy_request(state, parts, body_bytes, false).await
+            }
+        }
+    }
+}
+
+async fn proxy_request(
+    state: AppState,
+    parts: Parts,
+    body_bytes: Bytes,
+    use_execution_peer: bool,
+) -> Result<Response, (StatusCode, String)> {
+    let metrics = &state.metrics;
+    let peer_type = if use_execution_peer {
+        "execution"
+    } else {
+        "read"
+    };
+
+    let timer_histogram = metrics.request_latency.with_label_values(&[peer_type]);
+    let _timer = timer_histogram.start_timer();
+
+    metrics
+        .request_size_bytes
+        .with_label_values(&[peer_type])
+        .observe(body_bytes.len() as f64);
+
+    let peer_config = if use_execution_peer {
+        &state.execution_peer
+    } else {
+        &state.read_peer
+    };
+
+    let target_url = peer_config.address.clone();
+    // remove host header to avoid interfering with reqwest auto-host header
```

Comment:
can you expand on this? ordinarily a proxy would preserve the host header set by the client